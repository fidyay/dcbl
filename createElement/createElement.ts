import ClassComponent from "../ClassComponent/ClassComponent";
export interface ParametersProps {
  [prop: string]: unknown;
}

interface ElementProps {
  children: Child[];
  [prop: string]: unknown;
}

export interface DecibelElement {
  type: string;
  props: ElementProps;
}

type Child = DecibelElement | string;

const createElement = (
  type: string | (new (props: ElementProps) => ClassComponent),
  props: ParametersProps | null,
  ...children: Child[]
): DecibelElement | ClassComponent => {
  const elementProps = props
    ? { ...props, children }
    : ({ children } as ElementProps);
  if (typeof type === "string") {
    return {
      type,
      props: elementProps
    };
  }
  return new type(elementProps);
};

export default createElement;
