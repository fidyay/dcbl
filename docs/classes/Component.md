[dcbl](../README.md) / [Exports](../modules.md) / Component

# Class: Component<P, S\>

The base class for components of UI.

## Type parameters

| Name | Type |
| :------ | :------ |
| `P` | extends `BasicPropsAndStateInterface` = `BasicPropsAndStateInterface` |
| `S` | extends `BasicPropsAndStateInterface` = `BasicPropsAndStateInterface` |

## Table of contents

### Constructors

- [constructor](Component.md#constructor)

### Properties

- [\_manager](Component.md#_manager)
- [props](Component.md#props)
- [state](Component.md#state)

### Accessors

- [manager](Component.md#manager)

### Methods

- [componentDidMount](Component.md#componentdidmount)
- [componentDidUpdate](Component.md#componentdidupdate)
- [componentWillUnmount](Component.md#componentwillunmount)
- [render](Component.md#render)
- [setState](Component.md#setstate)
- [shouldComponentUpdate](Component.md#shouldcomponentupdate)

## Constructors

### constructor

• **new Component**<`P`, `S`\>(`props`)

Creates a [Component](Component.md) instance, saves props to component and sets default state if other wasn't provided.

#### Type parameters

| Name | Type |
| :------ | :------ |
| `P` | extends `object` = `object` |
| `S` | extends `object` = `object` |

#### Parameters

| Name | Type | Description |
| :------ | :------ | :------ |
| `props` | `P` | props from parent. |

#### Defined in

[Components/Component.ts:23](https://github.com/fidyay/dcbl/blob/1a6081b/Components/Component.ts#L23)

## Properties

### \_manager

• `Private` `Optional` **\_manager**: [`ComponentManager`](ComponentManager.md)<[`Component`](Component.md)<`P`, `S`\>\>

Object of a [ComponentManager](ComponentManager.md) class, which controls the component. It is not expected to be used by the consumer of a library.

#### Defined in

[Components/Component.ts:31](https://github.com/fidyay/dcbl/blob/1a6081b/Components/Component.ts#L31)

___

### props

• **props**: `P`

Props of a component, component gets them from the parent element, component rerenders when there is a change in props.

#### Defined in

[Components/Component.ts:16](https://github.com/fidyay/dcbl/blob/1a6081b/Components/Component.ts#L16)

___

### state

• **state**: `S`

State of a component. Each component has unique state. When state changes, component rerenders.

#### Defined in

[Components/Component.ts:18](https://github.com/fidyay/dcbl/blob/1a6081b/Components/Component.ts#L18)

## Accessors

### manager

• `set` **manager**(`m`): `void`

Setter of a [_manager](Component.md#_manager) property. Virtual dom sets it automaticly

#### Parameters

| Name | Type |
| :------ | :------ |
| `m` | [`ComponentManager`](ComponentManager.md)<`this`\> |

#### Returns

`void`

#### Defined in

[Components/Component.ts:33](https://github.com/fidyay/dcbl/blob/1a6081b/Components/Component.ts#L33)

## Methods

### componentDidMount

▸ **componentDidMount**(): `void`

Runs after the UI was created. By default does nothing.

#### Returns

`void`

#### Defined in

[Components/Component.ts:56](https://github.com/fidyay/dcbl/blob/1a6081b/Components/Component.ts#L56)

___

### componentDidUpdate

▸ **componentDidUpdate**(): `void`

Runs after component rerenders. By default does nothing.

#### Returns

`void`

#### Defined in

[Components/Component.ts:78](https://github.com/fidyay/dcbl/blob/1a6081b/Components/Component.ts#L78)

___

### componentWillUnmount

▸ **componentWillUnmount**(): `void`

Runs when component is going to be deleted from the UI. By default does nothing.

#### Returns

`void`

#### Defined in

[Components/Component.ts:81](https://github.com/fidyay/dcbl/blob/1a6081b/Components/Component.ts#L81)

___

### render

▸ `Abstract` **render**(): `TreeType`

Method that return the component's element tree. User should implement the method himself.

#### Returns

`TreeType`

#### Defined in

[Components/Component.ts:53](https://github.com/fidyay/dcbl/blob/1a6081b/Components/Component.ts#L53)

___

### setState

▸ `Protected` **setState**(`newState`): `Promise`<`void`\>

Sets new state to component and then rerenders it. Resetting the method to other value will break the library.

#### Parameters

| Name | Type | Description |
| :------ | :------ | :------ |
| `newState` | `S` \| `settingStateFunction`<`S`, `P`\> | new state object or syncronous or asyncronous function that sets new state. |

#### Returns

`Promise`<`void`\>

#### Defined in

[Components/Component.ts:39](https://github.com/fidyay/dcbl/blob/1a6081b/Components/Component.ts#L39)

___

### shouldComponentUpdate

▸ **shouldComponentUpdate**(`oldProps`, `newProps`): `boolean`

Runs when virtual dom checks wether the component should rerender on props change. By default checks every field of an props object with Object.is method.

#### Parameters

| Name | Type | Description |
| :------ | :------ | :------ |
| `oldProps` | `P` | old props object. |
| `newProps` | `P` | new props object, it will be set to the [props](Component.md#props) property. |

#### Returns

`boolean`

- true if UI should update, false otherwise.

#### Defined in

[Components/Component.ts:63](https://github.com/fidyay/dcbl/blob/1a6081b/Components/Component.ts#L63)
