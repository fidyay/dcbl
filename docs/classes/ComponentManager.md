[dcbl](../README.md) / [Exports](../modules.md) / ComponentManager

# Class: ComponentManager<C\>

Controls one of the [Component's](Component.md) instances. It is not intended to be used outside of element trees.

## Type parameters

| Name | Type |
| :------ | :------ |
| `C` | extends [`Component`](Component.md)<`any`, `any`\> |

## Table of contents

### Constructors

- [constructor](ComponentManager.md#constructor)

### Properties

- [VD](ComponentManager.md#vd)
- [component](ComponentManager.md#component)
- [componentChildTree](ComponentManager.md#componentchildtree)
- [indexInParent](ComponentManager.md#indexinparent)
- [parentNode](ComponentManager.md#parentnode)

### Methods

- [rerenderComponent](ComponentManager.md#rerendercomponent)

## Constructors

### constructor

• **new ComponentManager**<`C`\>(`component`)

Creates the instance of [ComponentManager](ComponentManager.md). Sets the component instance to [component](ComponentManager.md#component) field.
And also sets the created object to the [component's manager property](Component.md#_manager) using the setter.

#### Type parameters

| Name | Type |
| :------ | :------ |
| `C` | extends [`Component`](Component.md)<`any`, `any`, `C`\> |

#### Parameters

| Name | Type | Description |
| :------ | :------ | :------ |
| `component` | `C` | the instance of [Component](Component.md). |

#### Defined in

[Components/ComponentManager.ts:22](https://github.com/fidyay/dcbl/blob/1a6081b/Components/ComponentManager.ts#L22)

## Properties

### VD

• **VD**: [`VirtualDOM`](VirtualDOM.md)

The [VirtualDOM's](VirtualDOM.md) instance. It is set automaticaly by virtual dom.

#### Defined in

[Components/ComponentManager.ts:14](https://github.com/fidyay/dcbl/blob/1a6081b/Components/ComponentManager.ts#L14)

___

### component

• **component**: `C`

The [Component's](Component.md) instance. It is set automaticaly when the [ComponentManager's](ComponentManager.md) instance is created.

#### Defined in

[Components/ComponentManager.ts:16](https://github.com/fidyay/dcbl/blob/1a6081b/Components/ComponentManager.ts#L16)

___

### componentChildTree

• **componentChildTree**: `TreeType`

The previous tree generated by [Component's](Component.md) instance. It is used by virtual dom to compare trees.

#### Defined in

[Components/ComponentManager.ts:10](https://github.com/fidyay/dcbl/blob/1a6081b/Components/ComponentManager.ts#L10)

___

### indexInParent

• **indexInParent**: `number`

Index in child nodes of parent DOM element. It is set automaticaly by virtual dom.

#### Defined in

[Components/ComponentManager.ts:8](https://github.com/fidyay/dcbl/blob/1a6081b/Components/ComponentManager.ts#L8)

___

### parentNode

• **parentNode**: `HTMLElement`

Parent DOM element. It is set automaticaly by virtual dom.

#### Defined in

[Components/ComponentManager.ts:12](https://github.com/fidyay/dcbl/blob/1a6081b/Components/ComponentManager.ts#L12)

## Methods

### rerenderComponent

▸ **rerenderComponent**(): `void`

Rerenders the component's UI. And after that call the [componentDidUpdate](Component.md#componentdidupdate) method.

#### Returns

`void`

#### Defined in

[Components/ComponentManager.ts:27](https://github.com/fidyay/dcbl/blob/1a6081b/Components/ComponentManager.ts#L27)
