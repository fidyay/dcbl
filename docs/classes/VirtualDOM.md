[dcbl](../README.md) / [Exports](../modules.md) / VirtualDOM

# Class: VirtualDOM

Class of the virtual dom.

## Table of contents

### Constructors

- [constructor](VirtualDOM.md#constructor)

### Properties

- [DOMRoot](VirtualDOM.md#domroot)

### Methods

- [changeTree](VirtualDOM.md#changetree)
- [checkProps](VirtualDOM.md#checkprops)
- [createTree](VirtualDOM.md#createtree)
- [createTreeFromRoot](VirtualDOM.md#createtreefromroot)
- [runComponentWillUnmount](VirtualDOM.md#runcomponentwillunmount)

## Constructors

### constructor

• **new VirtualDOM**()

## Properties

### DOMRoot

• **DOMRoot**: `HTMLElement`

The root of the UI controlled by instance of a [VirtualDOM](VirtualDOM.md) class. It is set when the virtual dom is created.

#### Defined in

[VirtualDOM/VirtualDOM.ts:20](https://github.com/fidyay/dcbl/blob/6d857d5/VirtualDOM/VirtualDOM.ts#L20)

## Methods

### changeTree

▸ **changeTree**(`oldTree`, `newTree`, `parent`, `childIndex?`): `void`

Changes the UI to the new tree.

#### Parameters

| Name | Type | Description |
| :------ | :------ | :------ |
| `oldTree` | `TreeType` | the old tree's element. |
| `newTree` | `TreeType` | the new tree's element. |
| `parent` | `HTMLElement` | the parent DOM element. |
| `childIndex?` | `number` | index in childNodes of parent DOM element. |

#### Returns

`void`

#### Defined in

[VirtualDOM/VirtualDOM.ts:123](https://github.com/fidyay/dcbl/blob/6d857d5/VirtualDOM/VirtualDOM.ts#L123)

___

### checkProps

▸ **checkProps**(`oldProps`, `newProps`, `oldElementTemplate?`): `boolean`

Checks props of an element. If element is an instance of [DOMElementTemplate](DOMElementTemplate.md) class, it changes the attributes of a DOM element.

#### Parameters

| Name | Type | Description |
| :------ | :------ | :------ |
| `oldProps` | `any` | old props object. |
| `newProps` | `any` | new props object. |
| `oldElementTemplate?` | [`DOMElementTemplate`](DOMElementTemplate.md)<keyof `HTMLElementTagNameMap`\> | optianal parameter of an old [DOMElementTemplate's](DOMElementTemplate.md) instance, which stores the link to DOM element. |

#### Returns

`boolean`

- true if props were changed, false otherwise.

#### Defined in

[VirtualDOM/VirtualDOM.ts:41](https://github.com/fidyay/dcbl/blob/6d857d5/VirtualDOM/VirtualDOM.ts#L41)

___

### createTree

▸ **createTree**(`child`, `parent?`, `childIndex?`): `void`

Creates new DOM element tree and pushes inside to parent DOM element.

#### Parameters

| Name | Type | Description |
| :------ | :------ | :------ |
| `child` | `TreeType` | the tree element. |
| `parent?` | `HTMLElement` | the parent element. If it's not set, the [DOMRoot](VirtualDOM.md#domroot) is used. |
| `childIndex?` | `number` | index in childNodes of parent element. |

#### Returns

`void`

#### Defined in

[VirtualDOM/VirtualDOM.ts:223](https://github.com/fidyay/dcbl/blob/6d857d5/VirtualDOM/VirtualDOM.ts#L223)

___

### createTreeFromRoot

▸ **createTreeFromRoot**(`rootElement`, `rootDOMElement?`): `void`

Creates virtual dom tree from given tree element.

#### Parameters

| Name | Type | Description |
| :------ | :------ | :------ |
| `rootElement` | `TreeType` | root element. |
| `rootDOMElement?` | `HTMLElement` | root DOM element, saves it to [DOMRoot](VirtualDOM.md#domroot). If it's not provided, creates div with an id attribute set to "decibel-root". |

#### Returns

`void`

#### Defined in

[VirtualDOM/VirtualDOM.ts:287](https://github.com/fidyay/dcbl/blob/6d857d5/VirtualDOM/VirtualDOM.ts#L287)

___

### runComponentWillUnmount

▸ **runComponentWillUnmount**(`tree`): `void`

Runs [componentWillUnmount](Component.md#componentwillunmount) methods of all components inside the tree.

#### Parameters

| Name | Type | Description |
| :------ | :------ | :------ |
| `tree` | `TreeType` | element tree. |

#### Returns

`void`

#### Defined in

[VirtualDOM/VirtualDOM.ts:24](https://github.com/fidyay/dcbl/blob/6d857d5/VirtualDOM/VirtualDOM.ts#L24)
