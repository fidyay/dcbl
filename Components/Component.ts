import { childrenType } from "../createElement/createElement";
import ComponentManager from "./ComponentManager";

export type BasicPropsAndStateInterface = object;

type settingStateFunction<S, P> = (state: S, props: P) => S | Promise<S>;

abstract class Component<
  P extends BasicPropsAndStateInterface = BasicPropsAndStateInterface,
  S extends BasicPropsAndStateInterface = BasicPropsAndStateInterface
> {
  props: P;
  state!: S;
  constructor(props: P) {
    this.props = props;
    if (!this.state) {
      const defaultState = {};
      this.state = defaultState as S;
    }
  }
  private _manager?: ComponentManager<this>;
  set manager(m: ComponentManager<this>) {
    this._manager = m;
  }
  protected async setState(newState: S | settingStateFunction<S, P>) {
    if (typeof newState === "object") {
      this.state = newState;
    } else if (typeof newState === "function") {
      const updatedState = newState(this.state, this.props);
      if (updatedState instanceof Promise) {
        this.state = await updatedState;
      } else {
        this.state = updatedState;
      }
    }
    this._manager?.rerenderComponent();
  }
  abstract render(): childrenType;
}
export default Component;
