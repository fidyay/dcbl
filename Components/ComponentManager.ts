import Component from "./Component";
import VirtualDOM from "../VirtualDOM/VirtualDOM";
import { childrenType } from "../createElement/createElement";
class ComponentManager<C extends Component<any, any>> {
  componentChildTree!: childrenType;
  VD!: VirtualDOM;
  component: C;
  constructor(component: C) {
    this.component = component;
    component.manager = this;
  }
  rerenderComponent() {
    const newTree = this.component.render();
    return;
  }
}

export default ComponentManager;
