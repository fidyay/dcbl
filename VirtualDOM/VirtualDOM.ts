import ComponentManager from "../Components/ComponentManager";
import DOMElementTemplate from "../DOMElementTemplate/DOMElementTemplate";
import { childrenType, DOMElementType } from "../createElement/createElement";
import Component from "../Components/Component";
import Optionate from "../typeManipulation/optionate";
import RemoveFields from "../typeManipulation/removeField";
import { Styles } from "../global";

type EventType = (this: Element, ev: Event) => any;
class VirtualDOM {
  DOMRoot!: HTMLElement;
  createTree(child: childrenType, parent?: HTMLElement) {
    if (!parent) parent = this.DOMRoot;
    if (typeof child === "string") {
      const textNode = document.createTextNode(child);
      parent.append(textNode);
    } else if (child instanceof DOMElementTemplate) {
      const el = document.createElement(child.type);
      const propsArr = Object.keys(child.props) as (keyof typeof child.props)[];
      for (const attr of propsArr) {
        if (attr.slice(0, 2) === "on") {
          const event = attr.slice(2);
          const eventListener = child.props[attr] as EventType;
          el.addEventListener(event, eventListener);
        } else if (attr === "children" && child.props.children) {
          const children = child.props.children.flat();
          for (const child of children) {
            this.createTree(child, el);
          }
        } else if (attr === "style") {
          const styles = child.props[attr] as Styles;
          const stylesKeys = Object.keys(styles) as (keyof Styles)[];
          for (const style of stylesKeys) {
            const elStyles = el.style as unknown as { [key: string]: string };
            elStyles[style] = styles[style] as string;
          }
        } else {
          el.setAttribute(attr, child.props[attr] as any);
        }
        parent.append(el);
        child.DOMEl = el;
      }
    } else if (child instanceof ComponentManager) {
      child.VD = this;
      const treeTemplate = child.component.render();
      this.createTree(treeTemplate, parent);
    }
  }
  createTreeFromRoot(
    rootElement:
      | string
      | DOMElementTemplate<DOMElementType>
      | ComponentManager<Component<any, any>>,
    rootDOMElement?: HTMLElement
  ) {
    if (!rootDOMElement) {
      rootDOMElement = document.createElement("div");
      rootDOMElement.id = "decibel-root";
      document.body.append(rootDOMElement);
    }
    this.DOMRoot = rootDOMElement;
    this.createTree(rootElement);
  }
}

export default VirtualDOM;
