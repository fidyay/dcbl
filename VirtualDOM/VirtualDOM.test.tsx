import VirtualDOM, { TreeType } from "./VirtualDOM";
import Component from "../Components/Component";
import createElement from "../createElement/createElement";
import ComponentManager from "../Components/ComponentManager";

class List extends Component {
  render() {
    return (
      <ul>
        {["1", "2", "3"].map((str) => (
          <li>{str}</li>
        ))}
      </ul>
    );
  }
}

class Button extends Component {
  props!: { onClickListener: () => void };
  constructor(props: { onClickListener: () => void }) {
    super(props);
  }
  render() {
    return (
      <button style={{ color: "#fff" }} onclick={this.props.onClickListener}>
        +1
      </button>
    );
  }
}

class Link extends Component<any, { visited: boolean }> {
  constructor() {
    super({});
    this.state = {
      visited: false
    };
  }
  render() {
    return (
      <a
        className={this.state.visited ? "link_visited" : "link_unvisited"}
        onclick={this.onClickListener}
        disabled={this.state.visited}
        style={{
          color: this.state.visited ? "#ccc" : "#fff",
          display: "block"
        }}
      >
        Link to somewhere
      </a>
    );
  }
  async onClickListener() {
    await this.setState({ visited: true });
  }
}

class Clicker extends Component {
  state: { clicks: number };
  constructor() {
    super({});
    this.state = { clicks: 0 };
  }
  render() {
    return (
      <div>
        Clicks: {String(this.state.clicks)}
        <Button onClickListener={this.onClickListener} />
      </div>
    );
  }
  async onClickListener() {
    await this.setState({ clicks: this.state.clicks + 1 });
  }
}

class Todo extends Component<{
  children: number;
  completeTodo: () => Promise<any>;
}> {
  constructor(props: { children: number; completeTodo: () => Promise<any> }) {
    super(props);
  }
  render() {
    return (
      <li onclick={this.props.completeTodo}>{String(this.props.children)}</li>
    );
  }
}

class TodoList extends Component<any, { todos: ComponentManager<Todo>[] }> {
  constructor() {
    super({});
    const todos: ComponentManager<Todo>[] = [];
    for (let i = 0; i < 4; i++) {
      const todo = (
        <Todo
          completeTodo={async () => {
            const newArr = [...this.state.todos];
            newArr.splice(i, 1);
            await this.setState({ todos: newArr });
          }}
        >
          {i}
        </Todo>
      ) as ComponentManager<Todo>;
      todos.push(todo);
    }
    this.state = { todos };
  }
  render() {
    return <ul>{this.state.todos}</ul>;
  }
}

class ClickToWin extends Component<any, { won: boolean }> {
  constructor() {
    super({});
    this.state.won = false;
  }
  render() {
    return (
      <div>
        <header>Click to win</header>
        {this.state.won ? (
          <span>You won</span>
        ) : (
          <button onclick={this.win}>Win!</button>
        )}
        <footer>Lorem ipsum</footer>
      </div>
    );
  }
  async win() {
    await this.setState({ won: true });
  }
}

const link = <a href="https://google.com">Google</a>;

const textNode = "sample text";

describe("testing tree creation", () => {
  let vd: VirtualDOM;
  beforeEach(() => (vd = new VirtualDOM()));
  afterEach(() => (document.body.innerHTML = ""));
  test("creates text node", () => {
    vd.createTreeFromRoot(textNode);
    expect(document.body).toMatchSnapshot();
  });
  test("creates anchor element", () => {
    vd.createTreeFromRoot(link);
    expect(document.body).toMatchSnapshot();
  });
  test("inserts component tree", () => {
    vd.createTreeFromRoot(<Clicker />, document.body);
    expect(document.body).toMatchSnapshot();
  });
  test("inserts array of elements", () => {
    vd.createTreeFromRoot(<List />, document.body);
    expect(document.body).toMatchSnapshot();
  });
});

describe("testing tree changing", () => {
  let vd: VirtualDOM;
  beforeEach(() => (vd = new VirtualDOM()));
  afterEach(() => (document.body.innerHTML = ""));
  test("changes text in DOM", async () => {
    const cm = (<Clicker />) as ComponentManager<Clicker>;
    vd.createTreeFromRoot(cm);
    expect(document.body).toMatchSnapshot();
    await cm.component.onClickListener();
    expect(document.body).toMatchSnapshot();
    await cm.component.onClickListener();
    expect(document.body).toMatchSnapshot();
  });
  test("changes in props are checked", () => {
    const props1 = {
      a: 1,
      b: 2
    };
    const props2 = {
      a: 1,
      b: 3
    };
    expect(vd.checkProps(props1, props2)).toBe(true);
    const props3 = {
      a: 1,
      b: 2
    };
    const props4 = {
      a: 1,
      b: 2
    };
    expect(vd.checkProps(props3, props4)).toBe(false);
  });
  test("VD changes the attributes", async () => {
    const cm = (<Link />) as ComponentManager<Clicker>;
    vd.createTreeFromRoot(cm);
    expect(document.body).toMatchSnapshot();
    await cm.component.onClickListener();
    expect(document.body).toMatchSnapshot();
  });
  test("VD replaces the element", async () => {
    const cm = (<ClickToWin />) as ComponentManager<ClickToWin>;
    vd.createTreeFromRoot(cm);
    expect(document.body).toMatchSnapshot();
    await cm.component.win();
    expect(document.body).toMatchSnapshot();
  });
  test("VD deletes the elements", async () => {
    const cm = (<TodoList />) as ComponentManager<TodoList>;
    vd.createTreeFromRoot(cm);
    expect(document.body).toMatchSnapshot();
    const todoLast = cm.component.state.todos[3];
    await todoLast.component.props.completeTodo();
    expect(document.body).toMatchSnapshot();
    const todoSecond = cm.component.state.todos[1];
    await todoSecond.component.props.completeTodo();
    expect(document.body).toMatchSnapshot();
    const todoFirst = cm.component.state.todos[0];
    await todoFirst.component.props.completeTodo();
    expect(document.body).toMatchSnapshot();
  });
});
