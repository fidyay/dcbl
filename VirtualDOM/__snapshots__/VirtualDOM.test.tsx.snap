// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing tree changing VD changes the attributes 1`] = `
<body>
  <div
    id="decibel-root"
  >
    <a
      class="link_unvisited"
      disabled="false"
      style="color: rgb(255, 255, 255); display: block;"
    >
      Link to somewhere
    </a>
  </div>
</body>
`;

exports[`testing tree changing VD changes the attributes 2`] = `
<body>
  <div
    id="decibel-root"
  >
    <a
      class="link_visited"
      disabled="true"
      style="color: rgb(204, 204, 204); display: block;"
    >
      Link to somewhere
    </a>
  </div>
</body>
`;

exports[`testing tree changing VD deletes multiple elements 1`] = `
<body>
  <div
    id="decibel-root"
  >
    <div>
      <div>
        0
      </div>
      <div>
        1
      </div>
      <div>
        2
      </div>
      <div>
        3
      </div>
      <div>
        4
      </div>
      <div>
        5
      </div>
    </div>
  </div>
</body>
`;

exports[`testing tree changing VD deletes multiple elements 2`] = `
<body>
  <div
    id="decibel-root"
  >
    <div>
      <div>
        0
      </div>
      <div>
        1
      </div>
      <div>
        2
      </div>
    </div>
  </div>
</body>
`;

exports[`testing tree changing VD deletes the elements 1`] = `
<body>
  <div
    id="decibel-root"
  >
    <ul>
      <li>
        0
      </li>
      <li>
        1
      </li>
      <li>
        2
      </li>
      <li>
        3
      </li>
    </ul>
  </div>
</body>
`;

exports[`testing tree changing VD deletes the elements 2`] = `
<body>
  <div
    id="decibel-root"
  >
    <ul>
      <li>
        0
      </li>
      <li>
        1
      </li>
      <li>
        2
      </li>
    </ul>
  </div>
</body>
`;

exports[`testing tree changing VD deletes the elements 3`] = `
<body>
  <div
    id="decibel-root"
  >
    <ul>
      <li>
        0
      </li>
      <li>
        2
      </li>
    </ul>
  </div>
</body>
`;

exports[`testing tree changing VD deletes the elements 4`] = `
<body>
  <div
    id="decibel-root"
  >
    <ul>
      <li>
        2
      </li>
    </ul>
  </div>
</body>
`;

exports[`testing tree changing VD replaces the element 1`] = `
<body>
  <div
    id="decibel-root"
  >
    <div>
      <header>
        Click to win
      </header>
      <button>
        Win!
      </button>
      <footer>
        Lorem ipsum
      </footer>
    </div>
  </div>
</body>
`;

exports[`testing tree changing VD replaces the element 2`] = `
<body>
  <div
    id="decibel-root"
  >
    <div>
      <header>
        Click to win
      </header>
      <span>
        You won
      </span>
      <footer>
        Lorem ipsum
      </footer>
    </div>
  </div>
</body>
`;

exports[`testing tree changing changes text in DOM 1`] = `
<body>
  <div
    id="decibel-root"
  >
    <div>
      Clicks: 
      0
      <button
        style="color: rgb(255, 255, 255);"
      >
        +1
      </button>
    </div>
  </div>
</body>
`;

exports[`testing tree changing changes text in DOM 2`] = `
<body>
  <div
    id="decibel-root"
  >
    <div>
      Clicks: 
      1
      <button
        style="color: rgb(255, 255, 255);"
      >
        +1
      </button>
    </div>
  </div>
</body>
`;

exports[`testing tree changing changes text in DOM 3`] = `
<body>
  <div
    id="decibel-root"
  >
    <div>
      Clicks: 
      2
      <button
        style="color: rgb(255, 255, 255);"
      >
        +1
      </button>
    </div>
  </div>
</body>
`;

exports[`testing tree creation creates anchor element 1`] = `
<body>
  <div
    id="decibel-root"
  >
    <a
      href="https://google.com"
    >
      Google
    </a>
  </div>
</body>
`;

exports[`testing tree creation creates text node 1`] = `
<body>
  <div
    id="decibel-root"
  >
    sample text
  </div>
</body>
`;

exports[`testing tree creation inserts array of elements 1`] = `
<body>
  <ul>
    <li>
      1
    </li>
    <li>
      2
    </li>
    <li>
      3
    </li>
  </ul>
</body>
`;

exports[`testing tree creation inserts component tree 1`] = `
<body>
  <div>
    Clicks: 
    0
    <button
      style="color: rgb(255, 255, 255);"
    >
      +1
    </button>
  </div>
</body>
`;
